File: pipeline/analyze/test_cola_cloud.py
#!/usr/bin/env python3
"""
COLA Cloud API test -- explore data quality on free tier.
Uses 3-5 of our 500 monthly requests.

Usage:
    python -m pipeline.analyze.test_cola_cloud
"""

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

import httpx

from pipeline.lib.db import get_env

API_KEY = get_env("COLA_CLOUD_API_KEY")
BASE = "https://app.colacloud.us/api/v1"
DATA_DIR = Path(__file__).resolve().parents[2] / "data" / "imports"


def api_get(path: str, params: dict | None = None) -> dict:
    url = f"{BASE}{path}"
    print(f"  GET {path}")
    r = httpx.get(url, params=params or {}, headers={"X-API-Key": API_KEY}, timeout=30)
    r.raise_for_status()
    remaining = r.headers.get("x-ratelimit-remaining")
    if remaining:
        print(f"  -> Requests remaining: {remaining}")
    return r.json()


def main():
    print("\n=== COLA Cloud API Test ===\n")

    # Test 1: Total wine COLA count
    print("--- Test 1: Total wine COLA count ---")
    wines = api_get("/colas", {
        "product_type": "wine", "per_page": 1,
        "approval_date_from": "2015-01-01", "approval_date_to": "2026-12-31",
    })
    pagination = wines.get("pagination", {})
    print(f"Total wine COLAs: {pagination.get('total', 'unknown'):,}")
    print(f"Pages: {pagination.get('pages', 'unknown'):,}")

    # Test 2: Sample wine COLAs
    print("\n--- Test 2: Sample wine COLAs (100 records) ---")
    sample = api_get("/colas", {
        "product_type": "wine", "per_page": 100,
        "approval_date_from": "2025-01-01",
    })

    fields = {
        "brand_name": 0, "product_name": 0, "abv": 0,
        "wine_appellation": 0, "grape_varietals": 0,
        "wine_vintage_year": 0, "barcode_value": 0,
        "origin_name": 0, "class_name": 0,
    }
    for cola in sample.get("data", []):
        for field in fields:
            val = cola.get(field)
            if val is not None and val != "" and not (isinstance(val, list) and len(val) == 0):
                fields[field] += 1

    print("\nField coverage (out of 100):")
    for field, count in fields.items():
        print(f"  {field}: {count}%")

    print("\n--- Sample Records ---")
    for i, c in enumerate(sample.get("data", [])[:3]):
        print(f"\n[{i + 1}] {c.get('brand_name')} -- {c.get('product_name')}")
        print(f"    TTB ID: {c.get('ttb_id')}")
        print(f"    Class: {c.get('class_name')}")
        print(f"    Origin: {c.get('origin_name')} ({c.get('domestic_or_imported')})")
        print(f"    ABV: {c.get('abv')}")

    # Test 3: Detail record
    with_grapes = next(
        (c for c in sample.get("data", []) if c.get("grape_varietals") and len(c["grape_varietals"]) > 0),
        sample.get("data", [{}])[0] if sample.get("data") else {}
    )
    ttb_id = with_grapes.get("ttb_id")
    if ttb_id:
        print(f"\n--- Test 3: Detail for {ttb_id} ---")
        detail = api_get(f"/colas/{ttb_id}")
        d = detail.get("data", {})
        print(f"  Brand: {d.get('brand_name')}")
        print(f"  Product: {d.get('product_name')}")
        print(f"  Grapes: {json.dumps(d.get('grape_varietals'))}")
        print(f"  Appellation: {d.get('wine_appellation')}")
        print(f"  Barcode: {d.get('barcode_value')} ({d.get('barcode_type')})")

    # Test 4: Search for Ridge
    print('\n--- Test 4: Search for "Ridge" ---')
    ridge = api_get("/colas", {
        "product_type": "wine", "brand_name": "Ridge",
        "per_page": 10, "approval_date_from": "2015-01-01",
    })
    print(f"Ridge COLAs found: {ridge.get('pagination', {}).get('total')}")
    for c in ridge.get("data", [])[:5]:
        print(f"  {c.get('ttb_id')}: {c.get('brand_name')} -- {c.get('product_name')} ({c.get('abv')}% ABV)")

    # Test 5: Search for Opus One
    print('\n--- Test 5: Search for "Opus One" ---')
    opus = api_get("/colas", {
        "product_type": "wine", "q": "Opus One",
        "per_page": 10, "approval_date_from": "2015-01-01",
    })
    print(f"Opus One COLAs found: {opus.get('pagination', {}).get('total')}")
    for c in opus.get("data", [])[:5]:
        print(f"  {c.get('ttb_id')}: {c.get('brand_name')} -- {c.get('product_name')} | origin: {c.get('origin_name')}")

    # Save sample
    output = {
        "sample": sample.get("data", []),
        "ridge": ridge.get("data", []),
        "opus": opus.get("data", []),
        "detail": detail.get("data") if ttb_id else None,
        "total_wine_colas": pagination.get("total"),
    }
    out_path = DATA_DIR / "cola_cloud_sample.json"
    out_path.write_text(json.dumps(output, indent=2), encoding="utf-8")
    print(f"\nFull sample saved to {out_path}")
    print("Total API requests used: 5 of 500")


if __name__ == "__main__":
    main()

File: scripts_archive/node/scripts/test_cola_cloud.mjs
#!/usr/bin/env node
/**
 * COLA Cloud API test — explore data quality on free tier
 * Uses 3-5 of our 500 monthly requests
 */

import { readFileSync, writeFileSync } from 'fs';

// Read .env manually
const envContent = readFileSync('.env', 'utf8');
const API_KEY = envContent.match(/COLA_CLOUD_API_KEY=(.+)/)?.[1]?.trim();
const BASE = 'https://app.colacloud.us/api/v1';

async function apiGet(path, params = {}) {
  const url = new URL(`${BASE}${path}`);
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined) url.searchParams.set(k, v);
  }
  console.log(`  GET ${url.pathname}${url.search}`);
  const res = await fetch(url, {
    headers: { 'X-API-Key': API_KEY }
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`${res.status}: ${text}`);
  }
  const remaining = res.headers.get('x-ratelimit-remaining');
  if (remaining) console.log(`  → Requests remaining: ${remaining}`);
  return res.json();
}

async function main() {
  console.log('\n=== COLA Cloud API Test ===\n');

  // Test 1: Get total wine COLA count (1 request)
  console.log('--- Test 1: Total wine COLA count ---');
  const wines = await apiGet('/colas', {
    product_type: 'wine',
    per_page: 1,
    approval_date_from: '2015-01-01',
    approval_date_to: '2026-12-31'
  });
  console.log(`Total wine COLAs: ${wines.pagination?.total?.toLocaleString()}`);
  console.log(`Pages: ${wines.pagination?.pages?.toLocaleString()}`);

  // Test 2: Get a page of wine COLAs to see real data (1 request)
  console.log('\n--- Test 2: Sample wine COLAs (100 records) ---');
  const sample = await apiGet('/colas', {
    product_type: 'wine',
    per_page: 100,
    approval_date_from: '2025-01-01'
  });

  // Analyze field coverage
  const fields = {
    brand_name: 0, product_name: 0, abv: 0,
    wine_appellation: 0, grape_varietals: 0,
    wine_vintage_year: 0, barcode_value: 0, barcode_type: 0,
    llm_category: 0, llm_category_path: 0,
    origin_name: 0, class_name: 0,
    llm_tasting_note_flavors: 0, llm_wine_designation: 0,
    llm_product_description: 0
  };

  for (const cola of sample.data) {
    for (const field of Object.keys(fields)) {
      const val = cola[field];
      if (val !== null && val !== undefined && val !== '' &&
          !(Array.isArray(val) && val.length === 0)) {
        fields[field]++;
      }
    }
  }

  console.log('\nField coverage (out of 100):');
  for (const [field, count] of Object.entries(fields)) {
    console.log(`  ${field}: ${count}%`);
  }

  // Show 3 sample records in detail
  console.log('\n--- Sample Records ---');
  for (let i = 0; i < Math.min(3, sample.data.length); i++) {
    const c = sample.data[i];
    console.log(`\n[${i + 1}] ${c.brand_name} — ${c.product_name}`);
    console.log(`    TTB ID: ${c.ttb_id}`);
    console.log(`    Class: ${c.class_name}`);
    console.log(`    Origin: ${c.origin_name} (${c.domestic_or_imported})`);
    console.log(`    ABV: ${c.abv}`);
    console.log(`    Volume: ${c.volume} ${c.volume_unit}`);
    console.log(`    LLM Category: ${c.llm_category_path}`);
    console.log(`    Image: ${c.main_image_url ? 'yes' : 'no'}`);
  }

  // Test 3: Get a detail record to see wine-specific fields (1 request)
  // Pick first record with a grape varietal
  const withGrapes = sample.data.find(c =>
    c.grape_varietals && c.grape_varietals.length > 0
  ) || sample.data.find(c =>
    c.llm_category_path?.includes('Wine')
  ) || sample.data[0];

  console.log(`\n--- Test 3: Detail for ${withGrapes.ttb_id} ---`);
  const detail = await apiGet(`/colas/${withGrapes.ttb_id}`);
  const d = detail.data;
  console.log(`  Brand: ${d.brand_name}`);
  console.log(`  Product: ${d.product_name}`);
  console.log(`  Grapes: ${JSON.stringify(d.grape_varietals)}`);
  console.log(`  Appellation: ${d.wine_appellation}`);
  console.log(`  Vintage: ${d.wine_vintage_year}`);
  console.log(`  ABV: ${d.abv}`);
  console.log(`  Barcode: ${d.barcode_value} (${d.barcode_type})`);
  console.log(`  QR URL: ${d.qrcode_url}`);
  console.log(`  LLM Category: ${d.llm_category_path}`);
  console.log(`  LLM Description: ${d.llm_product_description}`);
  console.log(`  LLM Tasting Flavors: ${JSON.stringify(d.llm_tasting_note_flavors)}`);
  console.log(`  LLM Designation: ${d.llm_wine_designation}`);
  console.log(`  Permit: ${d.permit_number}`);
  console.log(`  Address: ${d.address_state} ${d.address_zip_code}`);
  console.log(`  Images: ${d.images?.length}`);
  console.log(`  Barcodes: ${JSON.stringify(d.barcodes)}`);

  // Test 4: Search for a known producer to test matching (1 request)
  console.log('\n--- Test 4: Search for "Ridge" ---');
  const ridge = await apiGet('/colas', {
    product_type: 'wine',
    brand_name: 'Ridge',
    per_page: 10,
    approval_date_from: '2015-01-01'
  });
  console.log(`Ridge COLAs found: ${ridge.pagination?.total}`);
  for (const c of ridge.data.slice(0, 5)) {
    console.log(`  ${c.ttb_id}: ${c.brand_name} — ${c.product_name} (${c.abv}% ABV)`);
  }

  // Test 5: Search for "Chateau Margaux" to test imported wines (1 request)
  console.log('\n--- Test 5: Search for "Opus One" ---');
  const opus = await apiGet('/colas', {
    product_type: 'wine',
    q: 'Opus One',
    per_page: 10,
    approval_date_from: '2015-01-01'
  });
  console.log(`Opus One COLAs found: ${opus.pagination?.total}`);
  for (const c of opus.data.slice(0, 5)) {
    console.log(`  ${c.ttb_id}: ${c.brand_name} — ${c.product_name} | origin: ${c.origin_name} | ABV: ${c.abv}`);
  }

  // Save full sample for offline analysis
  writeFileSync('data/imports/cola_cloud_sample.json', JSON.stringify({
    sample: sample.data,
    ridge: ridge.data,
    opus: opus.data,
    detail: detail.data,
    total_wine_colas: wines.pagination?.total
  }, null, 2));
  console.log('\n✅ Full sample saved to data/imports/cola_cloud_sample.json');
  console.log(`\nTotal API requests used: 5 of 500`);
}

main().catch(err => {
  console.error('Error:', err.message);
  process.exit(1);
});
